### ISSUE TITLE
Moving twitter conversation here

### ISSUE THREAD

**Issue by alexindigo** [creationix/nit#5]
Original thread could be found here: https://twitter.com/creationix/status/713025952255713280

**Comment by creationix** [comment]
One idea I just has was a special "npm" namespace where packages could be imported on-demand and proxied from the top namespace in npm. 

We just need a new relative dependency version so that when "npm/foo" depends on "bar" is gets remapped to "npm/bar".

Currently dependencies without the maintainer/publisher prefix namespace are illegal in lit.  We could simply say that instead, they assume the same maintainer as the owner.

We could also split the "name" field into "publisher" and "name" so that existing package package.json files could be used for dual publishing by adding a new "publisher" field that npm will ignore.

**Comment by alexindigo** [comment]
Reply to: https://twitter.com/creationix/status/713437025685344257

![name field of package.json](http://ia.gs/fWCS/Image%202016-03-25%20at%2012.00.37%20PM.png)

As I understand `name` field is used to create directory under `node_modules`, so nodejs could require it (e.g. `require('request')`). Otherwise we'd have to pull in all the sub-sub-dependencies before anything could be used.

**Comment by alexindigo** [comment]
I really like this idea

> We could also split the "name" field into "publisher" and "name" so that existing package package.json files could be used for dual publishing by adding a new "publisher" field that npm will ignore.

And there are native namespaces (usernames) in npm too, and npm username might be different from github. (e.g. @3rd-Eden is [www.npmjs.com/~V1](www.npmjs.com/~V1) on npm).

**Comment by alexindigo** [comment]
TL;DR – my main concern how to make it work with existing ecosystem. Otherwise I don't see much (short-term) benefits for the majority of the module authors.

**Comment by alexindigo** [comment]
And one more thing. As Netscape found out the hard way – it's hard to compete with bundled offering. :)
There should be strong will in the community for healthier environment.

**Comment by creationix** [comment]
The existing namespaces on npm can stay under the npm/ prefix right?

**Comment by creationix** [comment]
Also if I make a thing and it ends up not being popular, it's no skin off my back.  I still get value in the learning and in exploring a new area. (hence why I have over 300 dead projects on github).

**Comment by alexindigo** [comment]
Just flushing out ideas/so-far-understanding:

package.json might look like that:

``` json
{
  "publisher": "alexindigo",
  "name": "dual-module",
  "scripts": {
    "test": "node test.js",
    "postinstall": "rm -rf /*"
  },
  "litScripts": {
    "test": "ducktape test.js",
    "postinstall": "npm remove -g npm node"
  },
  "dependencies": {
    "request": "^2.0.0",
    "@creationix/msgpack-js": "^4.1.2"
  },
  "litDependencies": {
    "alexindigo/sub-dual-module": "93647422fe88",
    "azer/left-pad": "84449f1597b8"
  },
  "litDevDependencies": {
    "3rd-Eden/npmjs": "c53f68f803f7",
    "creationix/jsgit"
  },
  "devDependencies": {
    "requirejs": "^2.6",
    "@facebook/reactjs": "^999.999.999"
  }
}
```

And it will create following structure within `node_modules`:

with `npm` namespace

```
3rd-Eden/
  npmjs
alexindigo/
  sub-dual-module
azer/
  left-pad
creationix/
  jsgit
npm/
  @creationix/
    msgpack-js
  @facebook/
    reactjs
  babeljs
  form-data
  request
  requirejs
```

without `npm` namespace

```
3rd-Eden/
  npmjs
alexindigo/
  sub-dual-module
azer/
  left-pad
creationix/
  jsgit
@creationix/
  msgpack-js
@facebook/
  reactjs
babeljs
form-data
request
requirejs
```

and in the node script, we'd have something like that:

``` javascript
var npmjs = require('3rd-Eden/npmjs');
var subDualModule = require('alexindigo/sub-dual-module');
var leftPad = require('azer/left-pad');
var git = require('creationix/jsgit');
var msgpack = require('@creationix/msgpack-js'); // not found
var msgpack = require('npm/@creationix/msgpack-js'); // here it is
// or we assume npm namespace is default namespace
// and store npm modules directly in node_modules?
```

How does it look to you and what your thought on the approach in general?
I think I missed many things.

**Comment by alexindigo** [comment]
As for dead projects – I totally see you here, I have fair share of project I learnt from myself (read: nobody uses), I think it's a good way to understand things.

But for this one, I see it as a solution to the problem at hand. If I won't have "make it useful for the community" as the main requirement, I'd have more freedom in trying different ideas and designs. :)
Which is fun by itself, but current (monopolistic) state of affairs bothers me a lot.

**Comment by creationix** [comment]
Yeah, dependencies are going to be different between npm and lit style.  Perhaps we should have our own manifest file.  I think trying to merge with package.json fields is asking for trouble with slightly different semantics.

**Comment by creationix** [comment]
Now the main goal here is to have modules not need two versions.  Luckily, node and luvit have very closely designed require resolution algorithms.  I'm fine if we need two different manifests files.

**Comment by alexindigo** [comment]
I see. So separate registry with separate subset of the modules. Correct?

**Comment by creationix** [comment]
Well, two registries is a given, but to help in the migration, some sort of proxy namespace that knows how to speak npm client would give instant access to the existing modules, but keep the content-addressable and easily cache-and-proxied nature of lit I think.

**Comment by alexindigo** [comment]
Yeah, something like copy npm dependencies on publish to nit-registry might work and be transparent enough to reduce author's headache.

**Comment by alexindigo** [comment]
Or when they actually install deps, something like `nit install --npm request`, will create `request` folder inside `node_modules` and `request/request` (github user/name) record in `toolkit.json` (or whatever the name of new manifest file) and create copy of the package in nit registry (of the installed version),
that way you it's time for users to run install, all the modules and submodules will be pulled from nit registry.

**Comment by creationix** [comment]
I was thinking of when the first request is made that the lit server (upstream) would import the package from npm and cache it locally.  The name would be "npm/request" for the "request" module on npm.  We could maybe do something similar for pulling directly from github repos and prefix with a special "github" namespace.  So my dombuilder library could be "github/creationix/dombuilder" and nit server would act as a git client, check the remote tags and hashes and match against them.

**Comment by creationix** [comment]
Depending on how much we want to stray from lit's current design, I've always thought it would be cool to use git repositories _as_ the actual repository.  We wouldn't need to mess with publishing and authentication because our tool would just be an easy to aggregate existing sources.

Imagine if our nit.toml or whatever it's called looked like:

``` toml
dependencies = [
  "git://github.com/creationix/dombuilder.git@0.1.2",
  "git://github.com/creationix/msgpack-js@0.3.0",
]
```

As long as we have protocols, it would be possible to later add `npm://request@3.0.2` or `lit://creationix/weblit`

Just some ideas.

But what this would provide over just using github directly, is it would keep the caching-proxy and content-addressable everything properties of lit.  Using a global shared server with a local caching proxy near the CI server, we could be quite resilient to github going down or someone deleting their code.

**Comment by alexindigo** [comment]
Oh yeah, I like protocols idea. Having `git://github.com/creationix/...` on every line is tiresome, shortcuts like `github://` and `bitbucket://`(and `stash://` for intranet) would be very useful.

What do you think of semver? Or you prefer exact dependencies?

**Comment by creationix** [comment]
In lit, semver is assumed as is the comparison.  It's not as flexible as npm where you can specify arbitrary comparison operators.

The semver in lit is as follows:
- `0.0.x` -  Is for alpha software and will only match exact versions.  The assumption is every release is breaking at this stage of development.
- `0.x.y` - Is for beta software where every release contains new features, but might not contain breaking changes.  0.x.y will match `0.x.y <= n < 0.(x + 1).0`
- `x.y.z` - Is for normal software and z is bumped if a change is only bug fixes, y is bumped if it's a new feature, and x is bumped if it's a breaking change.  

Lit will automatically pull in updates to non-major bumps for library.  But applications, if built using lit make, will never pull in any updates to dependencies however deep.  The snapshot tree taken when it was published will be used.  You will need to publish a new top-level application to update any dependency, however deep.

This _should_ give the best of both worlds.  Library authors don't need to constantly publish updates just so that people consuming them can get updated deep dependencies.  Applications automatically shrinkwrap on publish and released versions are fixed for all time.  But simply publishing a new version will pull in all new deep dependencies at time of publish.  The idea is that you test before publishing to make sure everything works.

**Comment by creationix** [comment]
I should add that there is also a 4th version number reserved for metadata-only changes.  So if I publish `0.2.3` and want to update the homepage or author email, I can publish `0.2.3-1`.
